index.js

```diff
@@ -1,36 +1,44 @@
-var jsonMask = require('json-mask'),
-  compile = jsonMask.compile,
-  filter = jsonMask.filter;
+const jsonMask = require('json-mask')
+
+const badCode = code => code >= 300 || code < 200
 
 module.exports = function (opt) {
-  opt = opt || {};
+  opt = opt || {}
 
   function partialResponse(obj, fields) {
-    if (!fields) return obj;
-    return filter(obj, compile(fields));
+    if (!fields) return obj
+    return jsonMask(obj, fields)
   }
 
   function wrap(orig) {
     return function (obj) {
-      var param = this.req.query[opt.query || 'fields'];
+      const param = this.req.query[opt.query || 'fields']
       if (1 === arguments.length) {
-        orig(partialResponse(obj, param));
-      } else if (2 === arguments.length) {
-        if ('number' === typeof arguments[1]) {
-          orig(arguments[1], partialResponse(obj, param));
-        } else {
-          orig(obj, partialResponse(arguments[1], param));
-        }
+        return orig(partialResponse(obj, param))
+      }
+
+      if ('number' === typeof arguments[1] && !badCode(arguments[1])) {
+        // res.json(body, status) backwards compat
+        return orig(partialResponse(obj, param), arguments[1])
+      }
+
+      if ('number' === typeof obj && !badCode(obj)) {
+        // res.json(status, body) backwards compat
+        return orig(obj, partialResponse(arguments[1], param))
       }
-    };
+
+      // The original actually returns this.send(body)
+      return orig(obj, arguments[1])
+    }
   }
 
   return function (req, res, next) {
+    if (badCode(res.statusCode)) return next()
```

Comment:
What is the use-case for this check?

Reply:
As per your comment here https://github.com/nemtsov/express-partial-response/pull/4/files#r164286146
To not perform any modifications if a prior middleware already set a bad status.

On the other hand, the general idea behind that feature might be flawed.

On the one hand, if {errors: ['such and such']} is returned, that's good.
But if an object with over 9001 gorillion fields was returned, but the code was outside the range, that may be quite bad.

Reply:
Gotcha. It looks like I was wrong there. When line 36 is executed, the response shouldn't have a `statusCode` yet, since it runs on request and not after the response. I think your `badCode` checks in the wrap cover the cases for the two 2-argument cases, and you'd need to check `badCode(this.statusCode)` in the 1-argument case (since that happens before `res.json()` is called.

Reply:
Alright, I see I am missing cases if after the function is wrapped (while code was still good) someone sets status and then calls it with one arg.
Should I forego checking before wrap?

Reply:
> Alright, I see I am missing cases if after the function is wrapped (while code was still good) someone sets status and then calls it with one arg.

Exactly.

> Should I forego checking before wrap?

Yes. It's unnecessary, as far as I can tell.